extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -0,0 +1,371 @@
+     * element is the last result of `this.bindings_[i].expression`.
+     * @type {!Array<BindExpressionResultDef>}
+     */
+    this.previousResults_ = [];
+
+    /** @const {!Object} */
+    this.scope_ = Object.create(null);
+
+    /** @const {!../../../src/service/vsync-impl.Vsync} */
+    this.vsync_ = vsyncFor(ampdoc.win);
+
+    /** @const {!Function} */
+    this.boundMeasure_ = this.measure_.bind(this);
+
+    /** @const {!Function} */
+    this.boundMutate_ = this.mutate_.bind(this);
+
+    this.ampdoc.whenBodyAvailable().then(body => {
+      this.bindings_ = this.scanForBindings_(body);
+
+      // Trigger verify-only digest in development.
+      if (getMode().development) {
+        this.digest_(true);
+      }
+    });
+  }
+
+  /**
+   * Merges `state` into the current scope and immediately triggers a digest
+   * unless `opt_skipDigest` is false.
+   * @param {!Object} state
+   * @param {boolean=} opt_skipDigest
+   */
+  setState(state, opt_skipDigest) {
+    user().assert(this.enabled_, `Experiment "${TAG}" is disabled.`);
+
+    Object.assign(this.scope_, state);
+
+    if (!opt_skipDigest) {
+      this.digest_();
+    }
+  }
+
+  /**
+   * Scans children for attributes that conform to bind syntax and returns
+   * all bindings.
+   * @param {!Element} body
+   * @return {!Array<BindingDef>}
+   * @private
+   */
+  scanForBindings_(body) {
+    // TODO(choumx): Chunk if taking too long in a single frame.
+    const bindings = [];
+    const elements = body.getElementsByTagName('*');
+    for (let i = 0; i < elements.length; i++) {
+      const el = elements[i];
+      const attributes = el.attributes;
+      for (let j = 0; j < attributes.length; j++) {
+        const binding = this.bindingForAttribute_(attributes[j], el);
+        if (binding) {
+          bindings.push(binding);
+        }
+      }
+    }
+    return bindings;
+  }
+
+  /**
+   * Returns a struct representing the binding corresponding to the
+   * attribute param, if applicable.
+   * @param {!Attr} attribute
+   * @param {!Element} element
+   * @return {?BindingDef}
+   * @private
+   */
+  bindingForAttribute_(attribute, element) {
+    // TODO(choumx): Only allow binding to attributes allowed by validator.
+
+    const name = attribute.name;
+    if (name.length > 2) {
+      if (name[0] === '[' && name[name.length - 1] === ']') {
+        return {
+          property: name.substr(1, name.length - 2),
+          expression: attribute.value,
+          element,
+        };
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Schedules a vsync task to reevaluate all binding expressions.
+   * @param {boolean=} opt_verifyOnly
+   * @private
+   */
+  digest_(opt_verifyOnly) {
+    // TODO(choumx): Chunk if takes too long for a single frame.
+
+    /** {!VsyncTaskSpecDef} */
+    const task = {measure: this.boundMeasure_, mutate: this.boundMutate_};
+    this.vsync_.run(task, {
+      results: [],
+      verifyOnly: opt_verifyOnly,
```

Comment:
Done. This is only a problem if called by a class that has type-checking disabled, right?